117 eqjoin

adrienhaxaire opened the pull request:
Hi,

I've written two tests for eq-join, to contribute to the issue [117](https://www.rethinkdb.com/api/javascript/eq_join/).

I'm starting with Clojure, so let me know how I can improve my code! I have chosen to put these tests into a separate file because the fixtures are different, I need two tables to test the joins.

If it's OK, I'd like to add the tutorial examples from the [official website](https://www.rethinkdb.com/api/javascript/eq_join/), also for the inner and outer joins.

Thanks,
Adrien

danielcompton commented on test/rethinkdb/joins_test.clj:
```diff
@@ -0,0 +1,51 @@
+
+(defn reset [table data]
+  (with-open [conn (r/connect :db u/test-db)]
+    (-> (r/table-drop table)
```
Can we just delete everything from the table here instead?

adrienhaxaire replied:
OK

adrienhaxaire replied:
Done, see ae57cd09.

danielcompton commented on test/rethinkdb/joins_test.clj:
```diff
@@ -0,0 +1,51 @@
+            [rethinkdb.test-utils :as u]))
+
+(def names [{:id 1 :name "Pikachu", :type 2},
+            {:id 2 :name "Magnemite", :type [1 2]}])
```
Seeing Magnemite with two types here suggests that it will be somehow used in the joins below. It would be good to add a comment either here or on the joins, explaining why it doesn't show up.

adrienhaxaire replied:
That's also unclear to me. I thought that the eq-join would map the types in the array, but it doesn't. I can look into this in more details, just to confirm.

adrienhaxaire replied:
Done, see https://github.com/apa512/clj-rethinkdb/pull/175/commits/eb182465091b8e7005465a7b8eeb339e36f1ab1d.

I added a TODO in the commit to remind me of adding a test that maps the joins on the the types of Magnemite.

danielcompton commented on test/rethinkdb/joins_test.clj:
```diff
@@ -0,0 +1,51 @@
+(deftest eq-join
+  (testing "Simple eq-join"
+    (is (= 
+         (with-open [conn (r/connect :db u/test-db)]
```
can we use one connection for the whole thing?

adrienhaxaire replied:
Sure.

danielcompton commented on test/rethinkdb/joins_test.clj:
```diff
@@ -0,0 +1,51 @@
+        (r/run conn))))
+
+(defn setup-each [test-fn]
+  (reset :names names)
```
`setup-each` should create a `with-open` connection here like `setup-once` does, and pass it to `reset`

adrienhaxaire replied:
OK

danielcompton commented:
Hey, thanks for this! Happy for you to add the tutorial examples from the website for all the join types. Have added some comments inline to the code. Can you also squash all your commits into one?

adrienhaxaire commented:
You're welcome! Thank you for your fast answer and the detailed review. I am learning Clojure and RethinkDB, and writing tests is a nice way to experiment and learn.

I will apply your suggestions to have better code, then squash the commits (never done it, another thing to learn!) to have a clean merge. After that, shall I continue on this branch for the other tests, or should I create a new one?

danielcompton commented:
> After that, shall I continue on this branch for the other tests, or should I create a new one?

Probably easiest to make a new branch.

A good guide for squashing commits is at http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html.

Thanks!

adrienhaxaire commented:
Alright, did the code reviews, and squashed the commits into one. Thanks for the feedback. 

I think I messed up during my squashing, I pulled from this branch. The right commit is https://github.com/apa512/clj-rethinkdb/pull/175/commits/2fb7e667e31985f0442acee7fe5358cb16aa45cb. 

Please let me know if these modifications are okay and how I can fix this squashing thing :)

Oh and by the way, Circle CI complained about a failed test that I didn't touch in [build 513](https://circleci.com/gh/apa512/clj-rethinkdb/513). Did it occur before?

adrienhaxaire commented:
I saw that the reset function was still opening a connection, I fixed it as well. Squashed and pushed into https://github.com/apa512/clj-rethinkdb/pull/175/commits/e12f74b6d31454922370b208171edcb02c561db7. 

I had to pull to be able to push again. Next time, should I fast forward?

Sorry for being so noob with GitHub, thanks for your time!

dignati commented:
@adrienhaxaire I think you want to do `git push --force` after squashing the commits. This will overwrite the previous commits and also removes the need to merge the remote branch into your local one.

adrienhaxaire commented:
Thank you @dignati! I've reverted the merge then pushed with the `--force` option, seemed to work!

DonyorM commented:
@dignati is this good to be merged? I would like to build on what he's done for the tests for the other join functions.

DonyorM commented:
Is this ready to be merged? I would like to build on this pull request to write the tests for the other merge function.

apa512 commented:
I'm on my phone now but will check in a couple of hours.

apa512 commented:
@DonyorM merged
